delira/training/experiment.py

```diff
@@ -1,388 +1,582 @@
-
-
-from ..utils import now
-from ..data_loading import BaseDataManager, BaseLazyDataset
-from delira import __version__ as delira_version
-from .parameters import Parameters
-from ..models import AbstractNetwork
-from .abstract_trainer import AbstractNetworkTrainer
-from trixi.experiment import Experiment as TrixiExperiment
-import os
-import logging
 import typing
-import numpy as np
-
+import logging
 import pickle
-from abc import abstractmethod
-from sklearn.model_selection import KFold, StratifiedKFold, StratifiedShuffleSplit
-from sklearn.model_selection import train_test_split
+import os
 from datetime import datetime
-from inspect import signature
 from functools import partial
 
+import numpy as np
+from sklearn.model_selection import KFold, StratifiedKFold, \
+    StratifiedShuffleSplit, ShuffleSplit
+from trixi.experiment import Experiment as TrixiExperiment
+
 from delira import get_backends
 
+from ..data_loading import BaseDataManager
+from ..models import AbstractNetwork
+
+from .parameters import Parameters
+from .base_trainer import BaseNetworkTrainer
+from .predictor import Predictor
+
 logger = logging.getLogger(__name__)
 
 
-class AbstractExperiment(TrixiExperiment):
+class BaseExperiment(TrixiExperiment):
     """
-    Abstract Class Representing a single Experiment (must be subclassed for
-    each Backend)
+    Baseclass for Experiments.
+
+    Implements:
+
+    * Setup-Behavior for Models, Trainers and Predictors (depending on train
+        and test case)
 
-    See Also
-    --------
-    :class:`PyTorchExperiment`
+    * The K-Fold logic (including stratified and random splitting)
+
+    * Argument Handling
 
     """
-    @abstractmethod
-    def __init__(self, n_epochs, *args, **kwargs):
+
+    def __init__(self,
+                 params: Parameters,
```

Comment:
Is it possible to make the `params` object optional in the `__init__`? (It would change the signature and thus backwards compatibility could be problematic)

Reply:
We would have to change the argument order here and I'm not sure if I'd like that, because IMO, the parameters are more important than other stuff and I can't really imagine a scenario where you wouldn't pass parameters here. What do you think @ORippler?

Reply:
I also don't know of a situation where you wouldn't pass params to `__init__`. Though I think we should rethink on what to pass as additional kwarg and what to include inside the params (e.g. Everything that can be tuned/varied during training/testing should go into params, as they are intended to be utilized for the hyperparameter search)

Reply:
I encountered multiple situations where I wanted to initialise the experiment to get the `savepath` without initialising the parameters, but maybe that is only me ^^

Reply:
with current implementation, it must be typing.Union[str, Parameters]

Reply:
WIll fix this